[InstCombine] Simplify select if it combinated and/or/xor

ParkHanbum opened the pull request:
`and/or/xor` operations can each be changed to sum of logical
operations including operators other than themselves.

 `x&y -> (x|y) ^ (x^y)`
 `x|y -> (x&y) | (x^y)`
 `x^y -> (x|y) ^ (x&y)`

if left of condition of `SelectInst` is `and/or/xor` logical
operation and right is equal to `0, -1`, or a `constant`, and
if `TrueVal` consist of `and/or/xor` logical operation then we
can optimize this case.

This patch implements this combination.

Proof: https://alive2.llvm.org/ce/z/WW8iRR

Fixed https://github.com/llvm/llvm-project/issues/71792.

ParkHanbum committed:
[InstCombine] Simplify select if it combinated `and/or/xor` (#71792)

`and/or/xor` operations can each be changed to sum of logical
operations including operators other than themselves.

 `x&y -> (x|y) ^ (x^y)`
 `x|y -> (x&y) | (x^y)`
 `x^y -> (x|y) ^ (x&y)`

if left of condition of `SelectInst` is `and/or/xor` logical
operation and right is equal to `0, -1`, or a `constant`, and
if `TrueVal` consist of `and/or/xor` logical operation then we
can optimize this case.

This patch implements this combination.

Proof: https://alive2.llvm.org/ce/z/WW8iRR

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4572,12 +4614,37 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
 
+  if (ICmpInst::isEquality(Pred) && match(CmpRHS, m_APInt(C)) &&
+      (C->isZero() | C->isOne())) {
+    bool which;
```
nit: style (`Which`).

ParkHanbum replied:
after got a your review, I revised code overall. which is no need longer. thanks!

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4572,12 +4614,37 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
+    if (Pred == ICmpInst::ICMP_EQ)
+      which = C->isZero() ? false : true;
+    else
+      which = C->isZero() ? true : false;
```
`Which = Pred == ICmpInst::ICMP_EQ ? !C->isZero() : C->isZero()`

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4425,6 +4425,48 @@ Value *llvm::simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
+    return TrueWhenUnset ? TrueVal : FalseVal;
+  } else if (match(FalseVal, m_Or(m_Specific(X), m_Specific(Y))) &&
+             match(TrueVal, m_Xor(m_Specific(X), m_Specific(Y)))) {
+    return TrueWhenUnset ? TrueVal : FalseVal;
```
all of these are commutative so should the `m_c_...`.

ParkHanbum replied:
did it! thanks!

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4425,6 +4425,48 @@ Value *llvm::simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
+  // (X | Y) == 0 ? X ^ Y : X & Y   --> X & Y
+  // (X | Y) == 1 ? X & Y : X ^ Y   --> X & Y
+  // (X | Y) == 1 ? X ^ Y : X & Y   --> X ^ Y
+  if (match(TrueVal, m_And(m_Specific(X), m_Specific(Y))) &&
```
Think you have an issue. You can end up matching `X & Y == 0 ? X & Y : X ^ Y` -> `X ^ Y` which isn't valid. You need to either pass  the matching compare binop to `simplifySelectBitTestSpec` or just do the cmp match in this function.

ParkHanbum replied:
has been revised. as you mentioned, it has been changed to check the condition more specifically.

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4425,6 +4425,48 @@ Value *llvm::simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
+  // (X & Y) == 0 ? X | Y : X ^ Y   --> X ^ Y
+  // (X & Y) == 0 ? X ^ Y : X | Y   --> X | Y
+  // (X & Y) == 1 ? X | Y : X ^ Y   --> X | Y
+  // (X & Y) == 1 ? X ^ Y : X | Y   --> X ^ Y
```
in general `== 1` is really `!= 0`.

ParkHanbum replied:
you right. did modified.

ParkHanbum committed:
[InstCombine] Add test for simplify select if it combinated and/or/xor

`and/or/xor` operations can each be changed to sum of logical
operations including operators other than themselves.

 `x&y -> (x|y) ^ (x^y)`
 `x|y -> (x&y) | (x^y)`
 `x^y -> (x|y) ^ (x&y)`

This commit contains to test these cases.

Proof: https://alive2.llvm.org/ce/z/WW8iRR

ParkHanbum commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4572,12 +4614,37 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
+
   // Canonicalize ne to eq predicate.
   if (Pred == ICmpInst::ICMP_NE) {
     Pred = ICmpInst::ICMP_EQ;
```
I did add tests. but I didn't understand what exactly you said. could you please describe it to me more detail?

goldsteinn replied:
Rebase so that the test commit is before the implementation commit.
That way we can see the test changes the implementation causes.

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4425,6 +4425,60 @@ Value *llvm::simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
+  // (X ^ Y) == 0 could be transformed to X == Y
+  // (X ^ Y) != 0 could be transformed to X != Y
+  if (match(CmpLHS, m_CombineOr(m_c_Xor(m_Value(X), m_Value(Y)), m_Value(X)))) {
+    if (Y && match(CmpRHS, m_Zero()) || match(CmpRHS, m_Value(Y))) {
```
Does `X ^ Y == Y` / `Y == 0` work? If not think you need to refactor to something like 


```
X = nullptr; Y = nullptr;
if (!(match(CmpRHS, m_Zero()) && match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) {
 X = CmpLHS;
 Y = CmpRHS;
}
```

ParkHanbum replied:
I want to handle "(X^Y)==0" and "X==Y". is my code not enough?

goldsteinn replied:
You code will also hande `X == 0` and `X^Y == Y` which isn't what you want I think.

ParkHanbum replied:
how about this?
```
  if ((!match(CmpRHS, m_Zero()) && match(CmpLHS, m_Value(X)) &&
       match(CmpRHS, m_Value(Y))) ||
      (match(CmpRHS, m_Zero()) &&
       match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))))) {
 ```

goldsteinn replied:
sure that looks correct.

ParkHanbum replied:
fixed it! thanks for advise!

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4425,6 +4425,60 @@ Value *llvm::simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
+  // (X ^ Y) != 0 ? X & Y : X | Y   --> X & Y
+  // (X ^ Y) == 0 could be transformed to X == Y
+  // (X ^ Y) != 0 could be transformed to X != Y
+  if (match(CmpLHS, m_CombineOr(m_c_Xor(m_Value(X), m_Value(Y)), m_Value(X)))) {
```
for the condition check you need need commutative. You only need commutative for the ones with `m_Specific`

ParkHanbum replied:
I got it. thanks to know

ParkHanbum commented on llvm/test/Transforms/InstSimplify/select.ll:
```diff
@@ -1652,3 +1652,149 @@ define i8 @select_xor_cmp_unmatched_operands(i8 %0, i8 %1, i8 %c) {
+  %and = and i32 %x, %y
+  %cond = select i1 %xor1, i32 %or, i32 %and
+  ret i32 %cond
+}
```
sure, I was worried about many tests would be not welcome. 
I will add tests to enough that nothing worry about it.  :smile:

ParkHanbum replied:
I added tests! :smile:

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4434,6 +4434,61 @@ Value *llvm::simplifyWithOpReplaced(Value *V, Value *Op, Value *RepOp,
+  // (X & Y) == 0 ? X ^ Y : X | Y   --> X | Y
+  // (X & Y) != 0 ? X | Y : X ^ Y   --> X | Y
+  // (X & Y) != 0 ? X ^ Y : X | Y   --> X ^ Y
+  if (match(CmpLHS, m_c_And(m_Value(X), m_Value(Y))) &&
```
Don't need commutative and here (In general you don't need it on a commutative binop unless there is an `m_Specific` or `m_Deferred`). Likewise below.

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4618,6 +4673,9 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
 
+  if (Value *V = simplifySelectBitTestSpec(CmpLHS, CmpRHS, TrueVal, FalseVal))
+    return V;
+
```
We should see test changes in this commit.

dtcxzyw replied:
Please follow @goldsteinn's comment to show test changes in the second commit.

ParkHanbum commented:
@goldsteinn 
ping

dtcxzyw commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4640,6 +4692,9 @@ static Value *simplifySelectWithICmpCond(Value *CondVal, Value *TrueVal,
     }
   }
 
+  if (Value *V = simplifySelectBitTestSpec(CmpLHS, CmpRHS, TrueVal, FalseVal))
```
You should implement this logic in `simplifySelectWithICmpEq` (rebasing may be required).
Please add a negative test:
```
define i32 @src_select_and_eq0_or_xor(i32 %x, i32 %y) {
 ; CHECK-LABEL: @src_select_and_eq0_or_xor(
 ; CHECK-NEXT:    [[XOR:%.*]] = xor i32 [[X:%.*]], [[Y:%.*]]
 ; CHECK-NEXT:    ret i32 [[XOR]]
 ;
   %and = and i32 %x, %y
   %and0 = icmp slt i32 %and, 0
   %xor = xor i32 %x, %y
   %or = or i32 %x, %y
   %cond = select i1 %and0, i32 %or, i32 %xor
   ret i32 %cond
 }
```

Handling `ICMP_NE` is unnecessary because we always canonicalize `A != 0 ? C : D` into `A == 0 ? D : C`. See the code above this line:
https://github.com/llvm/llvm-project/blob/2bf01d73f6ebca11f36c17a65b7a86109d44681e/llvm/lib/Analysis/InstructionSimplify.cpp#L4620-L4624

ParkHanbum commented:
@dtcxzyw I misunderstood what goldsteinn said. hope I got it right this time.

dtcxzyw commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4585,6 +4585,57 @@ static Value *simplifySelectWithFakeICmpEq(Value *CmpLHS, Value *CmpRHS,
+static Value *simplifySpecificBitOp(Value *CmpLHS, Value *CmpRHS,
+                                    Value *TrueVal, Value *FalseVal) {
+  Value *X, *Y;
+  // (X & Y) == 0 ? X | Y : X ^ Y   --> X ^ Y
+  // (X & Y) == 0 ? X ^ Y : X | Y   --> X | Y
+  // (X & Y) != 0 ? X | Y : X ^ Y   --> X | Y
+  // (X & Y) != 0 ? X ^ Y : X | Y   --> X ^ Y
```
```suggestion
  // (X & Y) == 0 ? X | Y : X ^ Y   --> X ^ Y
  // (X & Y) == 0 ? X ^ Y : X | Y   --> X | Y
```
Remove the misleading comments.

dtcxzyw commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4585,6 +4585,57 @@ static Value *simplifySelectWithFakeICmpEq(Value *CmpLHS, Value *CmpRHS,
 }
 
+/// Try to simplify the Select instruction consisting of add/or/xor.
+static Value *simplifySpecificBitOp(Value *CmpLHS, Value *CmpRHS,
```
The function name `simplifySpecificBitOp` is ambiguous.

dtcxzyw commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4585,6 +4585,57 @@ static Value *simplifySelectWithFakeICmpEq(Value *CmpLHS, Value *CmpRHS,
                                Pred == ICmpInst::ICMP_EQ);
 }
 
+/// Try to simplify the Select instruction consisting of add/or/xor.
```
```suggestion
/// Try to simplify the Select instruction consisting of and/or/xor.
```
Typo

dtcxzyw commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4585,6 +4585,57 @@ static Value *simplifySelectWithFakeICmpEq(Value *CmpLHS, Value *CmpRHS,
+
+  // (X ^ Y) == 0 could be transformed to X == Y
+  // (X ^ Y) != 0 could be transformed to X != Y
+  if (match(CmpLHS, m_Value(X)) && match(CmpRHS, m_Value(Y))) {
+    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y)))) ||
+        (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))))
+      return FalseVal;
+  }
```
```suggestion
    if ((match(TrueVal, m_c_And(m_Specific(CmpLHS), m_Specific(CmpRHS))) &&
         match(FalseVal, m_c_Or(m_Specific(CmpLHS), m_Specific(CmpRHS)))) ||
        (match(FalseVal, m_c_And(m_Specific(CmpLHS), m_Specific(CmpRHS))) &&
         match(TrueVal, m_c_Or(m_Specific(CmpLHS), m_Specific(CmpRHS)))))
      return FalseVal;
```
These matches are no-op.

dtcxzyw replied:
Did you forget to fix it?

ParkHanbum replied:
I will change it to use `CmpLHS CmpRHS` to `X Y` instead using `match`  
ps. now I can use code review LOL

dtcxzyw commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4600,6 +4651,9 @@ static Value *simplifySelectWithICmpEq(Value *CmpLHS, Value *CmpRHS,
                              /* DropFlags */ nullptr, MaxRecurse) == FalseVal)
     return FalseVal;
 
+  if (simplifySpecificBitOp(CmpLHS, CmpRHS, TrueVal, FalseVal))
+    return FalseVal;
```
```suggestion
  if (Value *V = simplifySpecificBitOp(CmpLHS, CmpRHS, TrueVal, FalseVal))
    return V;
```
It is unsafe since we may handle more patterns in the future...

ParkHanbum replied:
got it!

dtcxzyw commented on llvm/test/Transforms/InstSimplify/select.ll:
```diff
@@ -1752,3 +1752,496 @@ define <4 x i32> @select_vector_cmp_with_bitcasts(<2 x i64> %x, <4 x i32> %y) {
+  ret i32 %cond
+}
+; X==Y ? X|Y : X&Y  -> X|Y
+define i32 @src_select_x_eq_y_and_or(i32 %x, i32 %y) {
```
Could you please add tests with `X != Y` as well?
```
X!=Y ? X|Y : X&Y -> X|Y
X!=Y ? X|Y : X&Y -> X&Y
```

dtcxzyw commented on llvm/test/Transforms/InstSimplify/select.ll:
```diff
@@ -1752,3 +1752,555 @@ define <4 x i32> @select_vector_cmp_with_bitcasts(<2 x i64> %x, <4 x i32> %y) {
+  ret i32 %cond
+}
+
+; X==Y ? X|Y : X&Y  -> X|Y
```
```suggestion
; X==Y ? X&Y : X|Y  -> X|Y
```

ParkHanbum replied:
my miss... sorry!

dtcxzyw commented on llvm/test/Transforms/InstSimplify/select.ll:
```diff
@@ -1752,3 +1752,555 @@ define <4 x i32> @select_vector_cmp_with_bitcasts(<2 x i64> %x, <4 x i32> %y) {
+  ret i32 %cond
+}
+
+; X!=Y ? X|Y : X&Y  -> X&Y
```
```suggestion
; X!=Y ? X|Y : X&Y  -> X|Y
```

ParkHanbum replied:
my miss... sorry!

dtcxzyw commented on llvm/test/Transforms/InstSimplify/select.ll:
```diff
@@ -1752,3 +1752,555 @@ define <4 x i32> @select_vector_cmp_with_bitcasts(<2 x i64> %x, <4 x i32> %y) {
+  ret i32 %cond
+}
+
+; X!=Y ? X|Y : X&Y  -> X|Y
```
```suggestion
; X!=Y ? X&Y : X|Y  -> X&Y
```

ParkHanbum replied:
my miss... sorry!

dtcxzyw reviewed (approved):
LGTM. Please wait for additional approval from other reviewers.

ParkHanbum commented:
@dtcxzyw I commented on the comment above, can you see it? It says (pending) on my screen, but I don't know why, but it may mean that it hasn't been uploaded. So I'll reply here again.



I think it is necessary. because Select ICmp is must be "X (comp) Y". here is example:
```
define i32 @src_select_xor_max_positive_int(i32 %x, i32 %y) {
; CHECK-LABEL: @src_select_xor_max_positive_int(
; CHECK-NEXT:    [[XOR:%.*]] = xor i32 [[X:%.*]], [[Y:%.*]]
; CHECK-NEXT:    [[XOR0:%.*]] = icmp eq i32 [[XOR]], 2147483647
; CHECK-NEXT:    [[AND:%.*]] = and i32 [[X]], [[Y]]
; CHECK-NEXT:    [[OR:%.*]] = or i32 [[X]], [[Y]]
; CHECK-NEXT:    [[COND:%.*]] = select i1 [[XOR0]], i32 [[AND]], i32 [[OR]]
; CHECK-NEXT:    ret i32 [[COND]]
;
  %xor = xor i32 %x, %y
  %xor0 = icmp eq i32 %xor, 2147483647
  %and = and i32 %x, %y
  %or = or i32 %x, %y
  %cond = select i1 %xor0, i32 %and, i32 %or
  ret i32 %cond
}
```
without `if (match(CmpLHS, m_Value(X)) && match(CmpRHS, m_Value(Y))) {` condition then
  `%and = and i32 %x, %y %or = or i32 %x, %y`
will match case.

am I right?

dtcxzyw commented:
> I commented on the comment above, can you see it? It says (pending) on my screen, but I don't know why, but it may mean that it hasn't been uploaded.

I guess you forgot to click **Review changes**->**Submit review** above the changed code :)

ParkHanbum commented:
> > I commented on the comment above, can you see it? It says (pending) on my screen, but I don't know why, but it may mean that it hasn't been uploaded.
> 
> I guess you forgot to click **Review changes**->**Submit review** above the changed code :)

my shame. I didn't know that!

dtcxzyw commented:
> @dtcxzyw I commented on the comment above, can you see it? It says (pending) on my screen, but I don't know why, but it may mean that it hasn't been uploaded. So I'll reply here again.
> 
> I think it is necessary. because Select ICmp is must be "X (comp) Y". here is example:
> 
> ```
> define i32 @src_select_xor_max_positive_int(i32 %x, i32 %y) {
> ; CHECK-LABEL: @src_select_xor_max_positive_int(
> ; CHECK-NEXT:    [[XOR:%.*]] = xor i32 [[X:%.*]], [[Y:%.*]]
> ; CHECK-NEXT:    [[XOR0:%.*]] = icmp eq i32 [[XOR]], 2147483647
> ; CHECK-NEXT:    [[AND:%.*]] = and i32 [[X]], [[Y]]
> ; CHECK-NEXT:    [[OR:%.*]] = or i32 [[X]], [[Y]]
> ; CHECK-NEXT:    [[COND:%.*]] = select i1 [[XOR0]], i32 [[AND]], i32 [[OR]]
> ; CHECK-NEXT:    ret i32 [[COND]]
> ;
>   %xor = xor i32 %x, %y
>   %xor0 = icmp eq i32 %xor, 2147483647
>   %and = and i32 %x, %y
>   %or = or i32 %x, %y
>   %cond = select i1 %xor0, i32 %and, i32 %or
>   ret i32 %cond
> }
> ```
> 
> without `if (match(CmpLHS, m_Value(X)) && match(CmpRHS, m_Value(Y))) {` condition then `%and = and i32 %x, %y %or = or i32 %x, %y` will match case.
> 
> am I right?

`match(CmpLHS, m_Value(X))` is equivalent to the following code :(
```
bool match(...) {
  X = CmpLHS;
  return true;
}
```

ParkHanbum commented:
> > @dtcxzyw I commented on the comment above, can you see it? It says (pending) on my screen, but I don't know why, but it may mean that it hasn't been uploaded. So I'll reply here again.
> > I think it is necessary. because Select ICmp is must be "X (comp) Y". here is example:
> > ```
> > define i32 @src_select_xor_max_positive_int(i32 %x, i32 %y) {
> > ; CHECK-LABEL: @src_select_xor_max_positive_int(
> > ; CHECK-NEXT:    [[XOR:%.*]] = xor i32 [[X:%.*]], [[Y:%.*]]
> > ; CHECK-NEXT:    [[XOR0:%.*]] = icmp eq i32 [[XOR]], 2147483647
> > ; CHECK-NEXT:    [[AND:%.*]] = and i32 [[X]], [[Y]]
> > ; CHECK-NEXT:    [[OR:%.*]] = or i32 [[X]], [[Y]]
> > ; CHECK-NEXT:    [[COND:%.*]] = select i1 [[XOR0]], i32 [[AND]], i32 [[OR]]
> > ; CHECK-NEXT:    ret i32 [[COND]]
> > ;
> >   %xor = xor i32 %x, %y
> >   %xor0 = icmp eq i32 %xor, 2147483647
> >   %and = and i32 %x, %y
> >   %or = or i32 %x, %y
> >   %cond = select i1 %xor0, i32 %and, i32 %or
> >   ret i32 %cond
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > without `if (match(CmpLHS, m_Value(X)) && match(CmpRHS, m_Value(Y))) {` condition then `%and = and i32 %x, %y %or = or i32 %x, %y` will match case.
> > am I right?
> 
> `match(CmpLHS, m_Value(X))` is equivalent to the following code :(
> 
> ```
> bool match(...) {
>   X = CmpLHS;
>   return true;
> }

Required to specify X and Y.

In the example above, CmpLHS is '%xor' and CmpRHS is '2147483647'. The operands of 'and' and 'or' are '%x, %y'.

`X==Y ? X&Y : I thought I needed `match(CmpLHS, m_Value(X)) && match(CmpRHS, m_Value(Y))` to check if X|Y` matches X|Y`.

then can check whether the operands of `icmp eq`, `and`, and `or` are all the same as `%x, %y`.


Yes, it is.

dtcxzyw reviewed (approved):
Still LGTM.

dtcxzyw commented:
Ping @nikic @goldsteinn

goldsteinn commented on llvm/lib/Analysis/InstructionSimplify.cpp:
```diff
@@ -4585,6 +4585,50 @@ static Value *simplifySelectWithFakeICmpEq(Value *CmpLHS, Value *CmpRHS,
+  }
+  // (X ^ Y) == 0 ? X | Y : X & Y   --> X & Y
+  // (X ^ Y) == 0 ? X & Y : X | Y   --> X | Y
+  if (match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) {
```
Does `X ^ Y == 0` ever not get canonicalized to `X == Y`?
If so I think this case is dead and you can just rely on the the below one (although you should keep the comments).

ParkHanbum replied:
When running with instcombine, xor is transformed to eq and comes. In case of instsimplify, it comes as xor.

If xor comes, is it desirable to process it by transform it to eq?

goldsteinn replied:
What you could also do to minimize the code duplication is do the following:

```
if ((match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) || (match(CmpLHS, m_Value(X)) && match(CmpRhs, m_Value(Y)))) {
```

Then drop the lines below.

@nikic, is there a general policy for whether we can rely on `InstCombine` canonicalizations in `InstSimplify`?

nikic replied:
InstSimplify should generally assume canonical form, unless there is some specific reason to the contrary.

ParkHanbum replied:
sure. however, I would like you to confirm that my understanding is correct.

"Change that remove X^Y transforming, leaving only X==Y, but retaining the comment." 

correct?

ParkHanbum replied:
@goldsteinn like this?
```
  // (X ^ Y) == 0 ? X | Y : X & Y   --> X & Y
  // (X ^ Y) == 0 ? X & Y : X | Y   --> X | Y
  // X == Y is equals with (X ^ Y) == 0
  LLVM_DEBUG(dbgs() << "xor\n");
  if ((match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) &&
       match(CmpRHS, m_Zero())) ||
      (match(CmpLHS, m_Value(X) && match(CmpRhs, m_Value(Y))))) {
    LLVM_DEBUG(dbgs() << "xor inside\n");
    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
         match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y)))) ||
        (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
         match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))))
      return FalseVal;
  }
```

goldsteinn replied:
No need for the explciit xor == 0 match. just:

```
  // (X ^ Y) == 0 ? X | Y : X & Y   --> X & Y
  // (X ^ Y) == 0 ? X & Y : X | Y   --> X | Y
 // X == Y is equals with (X ^ Y) == 0
  X = CmpLHS;
  Y = CmpRHS;
  if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
       match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y)))) ||
      (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
       match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))))
    return FalseVal;
```

is fine.

goldsteinn commented:
LGTM.

nikic commented:
A lot of these patterns look non-minimal to me.

For example, `(X | Y) == 0 ? (X & Y) : (X ^ Y)` to `X ^ Y` can instead be viewed as the combination of:

`(X | Y) == 0 ? (X & Y) : Z` to `(X | Y) == 0 ? 0 : Z` (https://alive2.llvm.org/ce/z/2ZJziy)

`(X | Y) == 0 ? 0 : (X ^ Y)` to `X ^ Y` (https://alive2.llvm.org/ce/z/_smtCP)

ParkHanbum commented:
@nikic I agree. while making the patch, I thought that I could make it more general-purpose using bit operations. However, due to lack of skills, I was unable to implement it. should I do pull request it later after implement better than.

ParkHanbum commented:
I think this simplification could be changed to:

The X & Y operation is an operator that activates the common bit between the two. Therefore, X & Y == 0 means there are no common bits between the two, ~X belongs to Y, and ~Y belongs to X.

This can be implied and simplified as follows:
```
int src(int x, int y, int c) {
   if (x&y == 0) return (x&c)&y;
   return c;
}

int tgt(int x, int y, int c) {
   if (x&y == 0) return y;
   return c;
}
```
Looking at this in IR, it looks like this.

1. Compiled Basic IR
```
define dso_local i32 @src_curr(i32 noundef %x, i32 noundef %y, i32 noundef %c) #0 {
entry:
   %cmp = icmp eq i32 %y, 0
   %conv = zext i1 %cmp to i32
   %and = and i32 %x, %conv
   %tobool = icmp ne i32 %and, 0
   br i1 %tobool, label %if.then, label %if.end

if.then: ; preds = %entry
   %and1 = and i32 %x, %c
   %and2 = and i32 %and1, %y
   br label %return

if.end: ; preds = %entry
   br label %return

return: ; preds = %if.end, %if.then
   %retval.0 = phi i32 [ %and2, %if.then ], [ %c, %if.end ]
   ret i32 %retval.0
}
```
2. IR after current optimization (optimization level 2-3)
```
define dso_local noundef i32 @tgt_curr(i32 noundef %x, i32 noundef %y, i32 noundef %c) local_unnamed_addr #0 {
entry:
   %cmp = icmp ne i32 %y, 0
   %0 = and i32 %x, 1
   %tobool.not1 = icmp eq i32 %0, 0
   %tobool.not = or i1 %cmp, %tobool.not1
   %and1 = and i32 %y, %x
   %and2 = select i1 %tobool.not, i32 -1, i32 %and1
   %retval.0 = and i32 %and2, %c
   ret i32 %retval.0
}
```
3. Simplify the basic IR as follows
```
define dso_local i32 @tgt_curr_opt(i32 noundef %x, i32 noundef %y, i32 noundef %c) #0 {
entry:
   %cmp = icmp eq i32 %y, 0
   %conv = zext i1 %cmp to i32
   %and = and i32 %x, %conv
   %tobool = icmp ne i32 %and, 0
   br i1 %tobool, label %if.then, label %if.end

if.then: ; preds = %entry
   br label %return

if.end: ; preds = %entry
   br label %return

return: ; preds = %if.end, %if.then
   %retval.0 = phi i32 [ %y, %if.then ], [ %c, %if.end ]
   ret i32 %retval.0
}
```
4. Optimize IR of 3 (Optimization Level 2)
```
define dso_local i32 @tgt_curr_opt(i32 noundef %x, i32 noundef %y, i32 noundef %c) #0 {
entry:
   %cmp = icmp ne i32 %y, 0
   %0 = and i32 %x, 1
   %tobool.not1 = icmp eq i32 %0, 0
   %tobool.not = or i1 %cmp, %tobool.not1
   %y.c = select i1 %tobool.not, i32 %c, i32 %y
   ret i32 %y.c
}
```
attaching the alive2 refer for this.
https://alive2.llvm.org/ce/z/vw2LhP

ParkHanbum commented:
I'm really sorry, but I've modified the code a lot. I would appreciate it if you could review again with be patient.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,110 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+    std::swap(TrueVal, FalseVal);
+  }
+
+  if (!isa<BinaryOperator>(CmpLHS) || !isa<BinaryOperator>(TrueVal))
```
i think more precise would be to match `m_BitwiseLogic`

ParkHanbum replied:
modified!

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,110 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+
+  // https://alive2.llvm.org/ce/z/EzU4sx
+  // (X & Y) == C ? X | Y : X ^ Y  ->  (X & Y) == C ?  C | (X^Y) : X ^ Y
+  // (X & Y) == C ? X ^ Y : X | Y  ->  (X & Y) == C ? ~C & (X|Y) : X | Y
```
The header comments (here and below) no longer match the impl.

Here for example it is:
`(X & Y) == C ? X | Y : (X ^ Y) | C -> (X ^ Y) | C`

ParkHanbum replied:
There was an error and it has been corrected.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,110 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+      if ((match(FalseVal, m_c_Xor(m_Specific(X), m_Specific(Y))) &&
+           match(CmpRHS, m_Zero())) ||
+          (match(FalseVal, m_c_Or(m_c_Xor(m_Specific(X), m_Specific(Y)),
+                                  m_Specific(CmpRHS)))))
```
Think a helper function would be useful here and below.

I.e instead of
```
(match(FalseVal, m_c_Xor(m_Specific(X), m_Specific(Y))) &&
           match(CmpRHS, m_Zero())) ||
          (match(FalseVal, m_c_Or(m_c_Xor(m_Specific(X), m_Specific(Y)),
                                  m_Specific(CmpRHS))))
```

you could do:

```
matchFalseVal = [FalseVal, CmpRHS, X, Y](unsigned OuterOpc, unsigned InnerOpc, bool NotRHS) {
   if(match(CmpRHS, m_Zero())
      return match(FalseVal, m_c_BinOp(InnerOpc, m_Specific(X), m_Specific(Y))`
   return match(FalseVal, m_c_BinOp(OuterOpc, m_c_BinOp(InnerOpc, m_Specific(X), m_Specific(Y)), NotRHS ? m_Not(m_Specific(CmpRHS)) : m_Specific(CmpRhs)));
};
```

ParkHanbum replied:
thanks! I reflected this, but since I couldn't use the ternary operator, I made it an if statement.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,115 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+                           unsigned OuterOpc, unsigned InnerOpc, bool NotRHS) {
+    if (match(CmpRHS, m_Zero()))
+      return match(FalseVal, m_c_BinOp(InnerOpc, m_Specific(X), m_Specific(Y)));
+
```
If `OuterOpc == NoOps` you should just return false here.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,115 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+    }
+  }
+  // https://alive2.llvm.org/ce/z/MnTYwj
+  // (X | Y) == C ? X & Y : FV  ->  (X | Y) == C ? C ^ (X^Y) : FV
```
FV is `X ^ Y` iirc

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,115 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+  }
+  // https://alive2.llvm.org/ce/z/MnTYwj
+  // (X | Y) == C ? X & Y : FV  ->  (X | Y) == C ? C ^ (X^Y) : FV
+  // (X | Y) == C ? X ^ Y : FV  ->  (X | Y) == C ? C ^ (X&Y) : FV
```
FV is `X & Y` iirc

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,115 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+               match(CmpRHS, m_Zero()))
+      return IC.replaceOperand(SI, 1, Null);
+  }
+  // https://alive2.llvm.org/ce/z/gSs5as
```
missing header comment with generalized pattern.

ParkHanbum replied:
you mean header comment need separate line between above?

goldsteinn replied:
As in comment with generalized transform (like you have for the above two cases).

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,138 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+      if (match(CmpRHS, m_AllOnes()))
+        return IC.replaceOperand(SI, 1, AllOnes);
+    }
+  }
```
This has become fairly complicate (in tracking all the conditions). You might try commenting the conditions with the case they represent to help keep track i.e for the xor case:

```
f (match(CmpLHS, m_Xor(m_Value(X), m_Value(Y)))) {
    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))))) {
         // (X ^ Y) == C ? X & Y : (X | Y) ^ C -> (X | Y) ^ C
      if (matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, OrOps, None) ||
         // (X ^ Y) == C ? X & Y : (X | Y) ^ C -> (X | Y) ^ C !! -- Same as above
          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, OrOps, None) ||
         // (X ^ Y) == C ? X ^ Y : (X | Y) & ~C -> (X | Y) & ~C
          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, OrOps, NotRHS))
        return IC.replaceInstUsesWith(*ISI, FalseVal);
    } else if (match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))) {
      // (X ^ Y) == C ? X | Y : (X & Y) | C -> (X & Y) | C
      if (matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, AndOps, None) ||
      // (X ^ Y) == C ? X | Y : (X & Y) | C -> (X & Y) | C !! -- Same as above
          matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, AndOps, None) ||
      // (X ^ Y) == C ? X | Y : (X & Y) ^ C -> (X & Y) ^ C
          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, AndOps, None))
        return IC.replaceInstUsesWith(*ISI, FalseVal);
      (X ^ Y) == -1 ? (X | Y) : FV -> (X ^ Y) == -1 ? -1 : FV
      if (match(CmpRHS, m_AllOnes()))
        return IC.replaceOperand(SI, 1, AllOnes);
    }
  }
```

ParkHanbum commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,138 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+    } else if (match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))) {
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, AndOps, None) ||
+          matchFalseVal(FalseVal, CmpRHS, X, Y, OrOps, AndOps, None) ||
```
sorry about that. fixed it.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,143 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+    if (match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))) {
+      // (X&Y)==0 ? X|Y : X^Y -> (X^Y)|0 : X^Y -> X^Y
+      if (matchFalseVal(FalseVal, CmpRHS, X, Y, NoOps, XorOps, None) ||
+          // (X&Y)==C ? X|Y : (X^Y)|C -> (X^Y)|C : (X^Y)|C -> (X|Y)^C
```
Think this is 
"// (X&Y)==C ? X|Y : (X^Y)|C -> (X^Y)|C : (X^Y)|C -> (X|Y)^C" -> "// (X&Y)==C ? X|Y : (X^Y)|C -> (X^Y)|C : (X^Y)|C -> (X^Y)^C"

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,143 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+      else if (match(CmpRHS, m_AllOnes()))
+        return IC.replaceOperand(SI, 1, AllOnes);
+    } else if (match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) {
+      // (X&Y)==0 ? X^Y : X|Y -> (X|Y)^0 : X|Y -> X:Y
```
Think comment should be:
"// (X&Y)==0 ? X^Y : X|Y -> (X|Y)^0 : X|Y -> X:Y" -> "// (X&Y)==0 ? X^Y : X|Y -> (X|Y)^0 : X|Y -> X|Y"

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,143 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+          // (X|Y)==C ? X^Y :~(X&Y)&C -> ~(X&Y)&C :~(X&Y)&C -> ~(X&Y)&C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, AndOps, NotInner))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+    } else if (match(TrueVal, m_BitwiseLogic(m_Specific(X), m_Specific(Y))) &&
```
You don't need to recheck that TrueVal is bitwise here (you check that above).

goldsteinn commented:
Aside from the few nit comments, this basically LGTM.

nikic commented:
@dtcxzyw Can you please re-test this?

dtcxzyw commented:
> @dtcxzyw Can you please re-test this?

Done.

dtcxzyw commented:
Ping. @nikic @goldsteinn

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,142 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+                 XorOps = Instruction::Xor, NoOps = 0;
+  enum NotMask { None = 0, NotInner, NotRHS };
+
+  auto matchFalseVal = [&](Value *FalseVal, Value *CmpRHS, Value *X, Value *Y,
```
I think the scoped `FalseVal` always the same as the passed in one?
In that case just use the capture `FalseVal`, shadow variables are confusing.

goldsteinn replied:
Likewise for X/Y

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,142 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+  // x^y -> (x|y) ^ (x&y)  or  (x|y) & ~(x&y)
+  Value *X, *Y;
+  if (!(match(CmpLHS, m_BitwiseLogic(m_Value(X), m_Value(Y)))) ||
+      !(match(TrueVal, m_BitwiseLogic(m_Specific(X), m_Specific(Y)))))
```
This needs to be `m_c_...` no?

ParkHanbum replied:
right. my mistake

goldsteinn replied:
See: https://github.com/llvm/llvm-project/pull/86632 you can rebase ontop of that I think it will get pushed quickly.

goldsteinn replied:
It has been pushed, so you can just rebase on main.

ParkHanbum replied:
@goldsteinn thanks. it's been reflected.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,142 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+      // (X&Y)==-1 ? X|Y : FV -> -1 : FV
+      else if (match(CmpRHS, m_AllOnes()))
+        return IC.replaceOperand(SI, 1, AllOnes);
```
This case could be higher up (i.e you don't need to match TrueVal to anything other than using either X and/or Y), really `X & Y == 1 ? AnyOp(X and/or Y)...` you can replace X/Y with -1.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,142 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, OrOps, NotRHS))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+      // (X&Y)==-1 ? X^Y : FV -> 0 : FV
+      else if (match(CmpRHS, m_AllOnes()))
```
See above

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,142 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+          // (X|Y)==C ? X^Y :~(X&Y)&C -> ~(X&Y)&C :~(X&Y)&C -> ~(X&Y)&C
+          matchFalseVal(FalseVal, CmpRHS, X, Y, AndOps, AndOps, NotInner))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+    } else if (match(CmpRHS, m_Zero()))
```
Same idea as the `X & Y == -1`, you can be constant folding `X -> 0` and `Y -> 0` for anything.

Maybe add a TODO to move these cases to `foldBinOpIntoSelect`?

ParkHanbum replied:
how about this?
is this implementation good enough to trying?

```
static Instruction *implyBitwiseCond(SelectInst &SI, Value *CmpLHS,
                                     Value *CmpRHS, Value *Val,
                                     InstCombinerImpl &IC) {
  LLVM_DEBUG(dbgs() << "ENTRY : "; CmpLHS->dump(); CmpRHS->dump());
  Instruction *I = dyn_cast<BinaryOperator>(Val);
  Instruction *RetI = I;
  bool Changed = false;
  for (int Idx = 0, Max = I->getNumOperands(); Idx < Max; Idx++) {
    Value *Op = I->getOperand(Idx);
    LLVM_DEBUG(dbgs() << "OPERAND : "; Op->dump());
    if (CmpLHS == Op) {
      RetI = IC.replaceOperand(*RetI, Idx, CmpRHS);
      Changed = true;
    }
    LLVM_DEBUG(dbgs() << "TTTT : "; I->dump());
  }

  LLVM_DEBUG(dbgs() << "TTTT : "; I->dump(); RetI->dump());
  return Changed ? I : 0;
}
```

```
  // (X&Y)==C ? X^Y : X|Y -> (X|Y)^C : X|Y  or (X|Y)&~C : X|Y
  if (match(CmpLHS, m_And(m_Value(X), m_Value(Y)))) {
    // (X&Y)==-1 ? X|Y : FV -> -1 : FV
    if (match(CmpRHS, m_AllOnes())) {
      if (auto I = implyBitwiseCond(SI, X, AllOnes, TrueVal, IC)) {
        if (auto II = implyBitwiseCond(SI, Y, AllOnes, I, IC))
          return IC.replaceOperand(SI, 1, II);
      }
    }
```

ParkHanbum replied:
here is results : 
```
IC: Visiting:   %cond = select i1 %cmp, i32 %or, i32 %xor
ENTRY : i32 %y
i32 -1
OPERAND : i32 %y
TTTT :   %or = or i32 -1, %x
OPERAND : i32 %x
TTTT :   %or = or i32 -1, %x
TTTT :   %or = or i32 -1, %x
  %or = or i32 -1, %x
ENTRY : i32 %x
i32 -1
OPERAND : i32 -1
TTTT :   %or = or i32 -1, %x
OPERAND : i32 %x
TTTT :   %or = or i32 -1, -1
TTTT :   %or = or i32 -1, -1
  %or = or i32 -1, -1
ADD DEFERRED:   %or = or i32 -1, -1
ADD DEFERRED:   %cond = select i1 %cmp, i32 %or, i32 %xor
IC: Mod =   %cond = select i1 %cmp, i32 %or, i32 %xor
    New =   %cond = select i1 %cmp, i32 %or, i32 %xor
ADD:   %cond = select i1 %cmp, i32 %or, i32 %xor
ADD:   %or = or i32 -1, -1
IC: Visiting:   %or = or i32 -1, -1
IC: Replacing   %or = or i32 -1, -1
    with i32 -1
IC: Mod =   %or = or i32 -1, -1
    New =   %or = or i32 -1, -1
IC: ERASE   %or = or i32 -1, -1
IC: Visiting:   %cond = select i1 %cmp, i32 -1, i32 %xor
```

goldsteinn replied:
I would say leave as todo in this patch. I think it better belongs more generically in `foldOpIntoSelect`.

ParkHanbum commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1672,6 +1672,142 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+          matchFalseVal(FalseVal, CmpRHS, X, Y, XorOps, AndOps, None))
+        return IC.replaceInstUsesWith(*ISI, FalseVal);
+      // (X^Y)==-1 ? (X|Y) : FV -> -1 : FV
+      if (match(CmpRHS, m_AllOnes()))
```
removed these kinds of combinations and labeled existing Tests with TODO.

goldsteinn commented:
LGTM.

ParkHanbum commented:
@goldsteinn thanks for review. I wish I could take TODOs.

goldsteinn commented:
> @goldsteinn thanks for review. I wish I could take TODOs.

You can, I think the way to do this is actually to handle this case in `computeKnownBitsFromCmp`.

We currently have:
```
 else if (match(LHS, m_And(m_V, m_APInt(Mask))) &&
               match(RHS, m_APInt(C))) {
      // For one bits in Mask, we can propagate bits from C to V.
      Known.Zero |= ~*C & *Mask;
      Known.One |= *C & *Mask;
      // assume(V | Mask = C)
    }
```

But we could (and should) also have:
```
 else if (match(LHS, m_c_And(m_V, m_Value()) &&
               match(RHS, m_APInt(C))) {
      Known.One |= *C;
   }
```

Likewise for the `or` case:

```
 else if (match(LHS, m_c_Or(m_V, m_Value()) &&
               match(RHS, m_APInt(C))) {
      Known.Zero |= ~*C;
   }
```

The only case that requires special logic is `xor` where we can't actually deduce any bits, but we
can deduce non-common bits.

That case I think you could handle here in a future patch.

Let me know if you are going to pursue a patch in `computeKnownBitsFromCmp`.

goldsteinn commented:
@dtcxzyw can you provide a second review of this?

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1687,6 +1687,120 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+
+  Value *AllOnes = Constant::getAllOnesValue(TrueVal->getType());
+  Value *Null = Constant::getNullValue(TrueVal->getType());
+  Instruction *ISI = &cast<Instruction>(SI);
```
```suggestion
```
You can use `&SI` directly.

ParkHanbum replied:
@dtcxzyw cannot be used as arugment for replaceInstUsesWith because SI is `SelectInst`. there's any way I don't know, please let me know.

dtcxzyw replied:
> @dtcxzyw cannot be used as arugment for replaceInstUsesWith because SI is `SelectInst`. there's any way I don't know, please let me know.

Can you try with `IC.replaceInstUsesWith(SI, Val)`?

ParkHanbum replied:
I'll try it and let you know

ParkHanbum replied:
it works! thanks!

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1687,6 +1687,120 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+  auto matchFalseVal = [&](unsigned OuterOpc, unsigned InnerOpc,
+                           unsigned NotMask) {
+    auto matchInner = m_c_BinOp(InnerOpc, m_Specific(X), m_Specific(Y));
+    if (OuterOpc == NoOps) {
+      if (match(CmpRHS, m_Zero()))
+        return match(FalseVal, matchInner);
+      else
+        return false;
+    }
```
```suggestion
    if (OuterOpc == NoOps) 
      return match(CmpRHS, m_Zero()) &&
        return match(FalseVal, matchInner);
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1687,6 +1687,120 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+
+  // Transform (X == C) ? X : Y -> (X == C) ? C : Y
+  // specific handling for Bitwise operation.
+  // https://alive2.llvm.org/ce/z/mW3eYR
```
It would be better to keep alive2 links in the PR description instead of codes/tests.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1687,6 +1687,120 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+  // x|y -> (x&y) | (x^y)  or  (x&y) ^  (x^y)
+  // x^y -> (x|y) ^ (x&y)  or  (x|y) & ~(x&y)
+  Value *X, *Y;
+  if (!(match(CmpLHS, m_BitwiseLogic(m_Value(X), m_Value(Y)))) ||
+      !(match(TrueVal, m_c_BitwiseLogic(m_Specific(X), m_Specific(Y)))))
```
```suggestion
  if (!match(CmpLHS, m_BitwiseLogic(m_Value(X), m_Value(Y))) ||
      !match(TrueVal, m_c_BitwiseLogic(m_Specific(X), m_Specific(Y))))
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1687,6 +1687,120 @@ static Value *foldSelectInstWithICmpConst(SelectInst &SI, ICmpInst *ICI,
+    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))))) {
+      // (X^Y)==C ? X&Y : (X|Y)^C -> (X|Y)^C
+      if (matchFalseVal(XorOps, OrOps, None) ||
+          // (X^Y)==C ? X&Y : (X|Y)&~C -> (X|Y)&~C
```
Can you hoisst the comment above the if?

ParkHanbum replied:
like this? 
```
// (X^Y)==C ? X&Y : (X|Y)^C -> (X|Y)^
// (X^Y)==C ? X&Y : (X|Y)&~C -> (X|Y)&~
if (matchFalseVal(XorOps, OrOps, None) ||
    matchFalseVal(AndOps, OrOps, NotRHS))
 ```

goldsteinn commented:
> > @goldsteinn thanks for review. I wish I could take TODOs.
> 
> You can, I think the way to do this is actually to handle this case in `computeKnownBitsFromCmp`.
> 
> We currently have:
> 
> ```
>  else if (match(LHS, m_And(m_V, m_APInt(Mask))) &&
>                match(RHS, m_APInt(C))) {
>       // For one bits in Mask, we can propagate bits from C to V.
>       Known.Zero |= ~*C & *Mask;
>       Known.One |= *C & *Mask;
>       // assume(V | Mask = C)
>     }
> ```
> 
> But we could (and should) also have:
> 
> ```
>  else if (match(LHS, m_c_And(m_V, m_Value()) &&
>                match(RHS, m_APInt(C))) {
>       Known.One |= *C;
>    }
> ```
> 
> Likewise for the `or` case:
> 
> ```
>  else if (match(LHS, m_c_Or(m_V, m_Value()) &&
>                match(RHS, m_APInt(C))) {
>       Known.Zero |= ~*C;
>    }
> ```
> 
> The only case that requires special logic is `xor` where we can't actually deduce any bits, but we can deduce non-common bits.
> 
> That case I think you could handle here in a future patch.
> 
> Let me know if you are going to pursue a patch in `computeKnownBitsFromCmp`.
> 
> Edit: Maybe wait a bit, I looked at this a bit and it creates some bugs related to #87015, so until that is resolved, not much use.

Think its a non-issue actually. @ParkHanbum do you want to make a patch for this?

dtcxzyw reviewed (approved):
LGTM. BTW the current implementation is a bit larger than the first version I approved. I am not sure whether it is an over-generalization. So I will leave the final decision to @goldsteinn and @nikic.

goldsteinn commented:
LGTM

goldsteinn commented:
@ParkHanbum do you have commit access or do you need me to merge for you?

ParkHanbum commented:
@goldsteinn I don't remember receiving any commit access. can I ask you why you asking that to me because I'm newbie.
